src/libraries/System.Text.RegularExpressions/src/System/Text/RegularExpressions/RegexCompiler.cs

```diff
@@ -793,6 +793,43 @@ private void CallToLower()
             }
         }
 
+        /// <summary>Gets whether the specified character participates in case conversion.</summary>
+        /// <remarks>
+        /// This method is used to perform operations as if they were case-sensitive even if they're
+        /// specified as being case-insensitive.  Such a reduction can be applied when the only character
+        /// that would lower-case to the one being searched for / compared against is that character itself.
+        /// </remarks>
+        private static bool ParticipatesInCaseConversion(int comparison)
+        {
+            Debug.Assert((uint)comparison <= char.MaxValue);
+
+            switch (char.GetUnicodeCategory((char)comparison))
+            {
+                case UnicodeCategory.ClosePunctuation:
+                case UnicodeCategory.ConnectorPunctuation:
+                case UnicodeCategory.Control:
+                case UnicodeCategory.DashPunctuation:
+                case UnicodeCategory.DecimalDigitNumber:
+                case UnicodeCategory.FinalQuotePunctuation:
+                case UnicodeCategory.InitialQuotePunctuation:
+                case UnicodeCategory.LineSeparator:
+                case UnicodeCategory.OpenPunctuation:
+                case UnicodeCategory.OtherNumber:
+                case UnicodeCategory.OtherPunctuation:
+                case UnicodeCategory.ParagraphSeparator:
+                case UnicodeCategory.SpaceSeparator:
+                    // All chars in these categories meet the criteria that the only way
```

Comment:
Is this always going to be the case forever, or is it possible a character could in future be added to one of these categories for which this did not hold true?

Reply:
Tarek and I consulted on this. The only categories that today change case are _*Letter_, _LetterNumber_, _OtherSymbol_, and _NonSpacingMark_. So we took what was left and further pared it down just to be safe. I'd be very surprised if future modifications to the standard would render this list incorrect.

Reply:
going through the whole Unicode data, I am seeing the only following category can have casing:

```
UppercaseLetter    Letter, Uppercase UppercaseLetter
LowercaseLetter     Letter, Lowercase
TitlecaseLetter        Letter, Titlecase
NonSpacingMark   Mark, Non-Spacing
LetterNumber        Number, Letter
OtherSymbol         Symbol, Other
```